### PR TITLE
Net autoinit

### PR DESCRIPTION
**Pull request by isage**
* Implement network auto-initialization on first call to net-related function
* Fix inet_ntop/inet_pton error handling (error handling changed due to scenet_errno returning SceNetKernelErrorCode instead of SceNetErrorCode)

Requires https://github.com/vitasdk/buildscripts/pull/112

### CODE REVIEW

**Comment by d3m3vilurr** [comment]
can you make gethostname and sethostname parts to different pr?

**Comment by isage** [comment]
I can, but it requires this branch due to call to net init

**Comment by d3m3vilurr** [comment]
no. i mean a patch should have just same purpose.
for example, this pr is `adding net init code into network functions` but `adding gethostname & sethostname` is outside of this purpose.
samely if pr has bugfix part in commits, they also split to an another pr.

if `adding gethostname & sethostname` is high priority, you should remove `netinit` part from their codesets, than add patch set in here.

so, you should change the pr to multiple cases;
1. split `get/sethostname function`
2. fix missing headers
3. inject netinit helper

and also i think i need a reason why you change the error handling part if you can. (imo maybe some functions will return the different error code. is it right?)

**Comment by isage** [comment]
Fine. I've removed everything, except netinit and inet_pton/ntop

File: newlib/libc/sys/vita/error.c [hunk 1]
```diff
@@ -45,174 +45,174 @@ int __vita_make_sce_errno(int posix_errno)
 
 int __vita_scenet_errno_to_errno(int sce_errno)
 {
-	switch (sce_errno)
+	switch (sce_errno & SCE_ERRNO_MASK)
 	{
-		case SCE_NET_ERROR_EPERM:
+		case SCE_NET_EPERM:
```
Comment (d3m3vilurr) [inline newlib/libc/sys/vita/error.c:50]: could you explain why did you change this?

File: newlib/libc/sys/vita/vitanet.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#ifndef SCE_NET_CTL_ERROR_NOT_TERMINATED
+#define SCE_NET_CTL_ERROR_NOT_TERMINATED 0x80412102
+#endif
+
+static void *_vita_net_memory = NULL;
+
+int _vita_net_init()
+{
+	int ret;
+	SceNetInitParam net_init_param = {0};
```
Comment (d3m3vilurr) [inline newlib/libc/sys/vita/vitanet.c:54]: maybe this approach is useful. but most projects already have network initialization code in their codeset.
also it makes useless calling when we use the network functions in every times.
so I'm worried it can make the perf problem to homebrews.

how about add a flag variable in top of the function to bypass below logics?

for example
```
static bool _VITA_NET_INITIALIZED = false;

int _vita_net_init()
{
  if (_VITA_NET_INITIALIZED)
    return 0;
  ...
  ret = secNetInit(...);
  ...
  if (ret != SCE_NET_ERROR_EBUSY)
    goto success;
  ...
success:
  _VITA_NET_... = true;
  return 0;
}
```

**Review by d3m3vilurr — CHANGES_REQUESTED** [review]

File: newlib/libc/sys/vita/error.c (hunk 1 above)
Comment (isage) [reply newlib/libc/sys/vita/error.c:50]: It's explained in pullrequest description

Comment (d3m3vilurr) [reply newlib/libc/sys/vita/error.c:50]: oh... got it... can you add the information into commit message?

Comment (isage) [reply newlib/libc/sys/vita/error.c:50]: Done

File: newlib/libc/sys/vita/vitanet.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	if (ret < 0)
+	{
+		if (ret != SCE_NET_CTL_ERROR_NOT_TERMINATED)
+		{
+			// something gone wrong, tear down completely, so it can be re-inited
+			// and we won't leak mem
+			sceNetTerm();
+			free(_vita_net_memory);
+		}
+		return ret;
```
Comment (d3m3vilurr) [inline newlib/libc/sys/vita/vitanet.c]: imo this line has to move to inside of `if` block

**Review by d3m3vilurr — CHANGES_REQUESTED** [review]

File: newlib/libc/sys/vita/vitanet.c (hunk 3 above)
Comment (isage) [reply newlib/libc/sys/vita/vitanet.c]: done

File: newlib/libc/sys/vita/vitanet.c [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+	ret = sceNetCtlInit();
+	if (ret < 0)
+	{
+		if (ret != SCE_NET_CTL_ERROR_NOT_TERMINATED)
+		{
+			// something gone wrong, tear down completely, so it can be re-inited
+			// and we won't leak mem
+			sceNetTerm();
+			free(_vita_net_memory);
```
Comment (d3m3vilurr) [inline newlib/libc/sys/vita/vitanet.c]: if `sceNetInit` is `SCE_NET_ERROR_EBUSY`. this code call `free` twice...

so imo, there are two options
1. just change the code like `sceNetInit` part. but remove `free`. imo we don't need `sceNetTerm()` too.
2. use 1 part and also use _vita_net_initialized more complexly.

```
enum {
  NOT_INIT,
  NET_MODULE_LOADED,
  NET_INITIALIZED,
  NETCTL_INITIALIZED,
};
...
static _vita_.. = NOT_INIT;
...

switch (_vita_...) {
  case NOT_INIT: {
    if (sceIsModule...
      ...
    }
    _vita_...++;
  }
  case NET_MODULE_LOADED: {
    ...
    _vita_...++;
  }
  case NET_INITIALIZED: {
    ret = sceNetCtlInit();
    if (ret < 0 && ret != SCE_NET_CTL_ERROR_NOT_TERMINATED) {
      return ret;
    }
    _vita_...++;
  }
  case NETCTL_INITIALIZED:
    return 0;
}
```

Comment (isage) [reply newlib/libc/sys/vita/vitanet.c]: Fixed. We'll just check if mem is allocated (= we inited Net module)

Comment (d3m3vilurr) [reply newlib/libc/sys/vita/vitanet.c]: ok. LGTM now. :+1:
